src/parser/jobs/mnk/Fists.js

```diff
@@ -28,28 +54,51 @@ export default class Fists extends Module {
 
 	static title = 'Fist Stances'
 
-	_stanceApplied = []
-	_stanceDropped = []
-	_stanceless = 0
+	_activeFist = STANCES.find(fist => this.combatants.selected.hasStatus(fist)) || STANCELESS
+	_fistUptime = {[STANCELESS]: 0} // Initialise stanceless to prevent weird UI shit
+	_fistGCDs = {[STANCELESS]: 0} // Initialise empty
+
+	_lastFistChange = this.parser.fight.start_time
 
 	constructor(...args) {
 		super(...args)
+		this.addHook('cast', {by: 'player'}, this._onCast)
 		this.addHook('applybuff', {to: 'player', abilityId: STANCES}, this._onGain)
 		this.addHook('removebuff', {to: 'player', abilityId: STANCES}, this._onRemove)
 		this.addHook('complete', this._onComplete)
 	}
 
-	_onGain(event) {
-		if (this._stanceDropped.length > 0) {
-			this._stanceless += event.timestamp - _.last(this._stanceDropped).timestamp
+	handleFistChange(stanceId) {
+		if (!this._fistUptime.hasOwnProperty(this._activeFist)) {
+			this._fistUptime[this._activeFist] = 0
 		}
 
-		this._stanceApplied.push(event)
-		return
+		this._fistUptime[this._activeFist] += (this.parser.currentTimestamp - this._lastFistChange)
+		this._lastFistChange = this.parser.currentTimestamp
+		this._activeFist = stanceId
 	}
 
-	_onRemove(event) {
-		this._stanceDropped.push(event)
+	_onCast(event) {
+		const action = getAction(event.ability.guid)
+
+		if (!action || !action.onGcd) {
+			return
+		}
+
+		this._fistGCDs[this._activeFist]++
+	}
+
+	_onGain(event) {
+		this.handleFistChange(event.ability.guid)
+	}
+
+	_onRemove() {
+		// This is kinda flaky, there's a chance for remove to run after gain and effectively
```

Comment:
I can almost guarantee it's the diff between

```
...
wind cast
fire remove
wind apply
...
```

and 

```
...
wind cast
wind apply
fire remove
...
```

To work around/with that, I'd say it's probably safe to ignore cases where the stance being removed != current active stance.